File: backend/app/wechat/service.py
# -*- coding: utf-8 -*-
"""微信登录、消息循环和 Agent 的组合服务。"""
from __future__ import annotations

import os
import asyncio
from datetime import date
from typing import Any

from ..agent.model_client import ModelError
from ..agent.runner import AgentRunner
from ..agent.session_store import SessionStore
from ..services import work_items
from .. import clock, db
from .auth_service import AuthService
from .config import public_config, save_config
from .credential_store import CredentialStore
from .ilink_client import ILinkClient, ILinkConfig, ILinkError
from .message_loop import MessageLoop
from .models import IncomingText


class WeChatService:
    def __init__(self):
        self.credentials = CredentialStore()
        self.client = ILinkClient(ILinkConfig(
            base_url=os.environ.get('MEIMEI_WECHAT_BASE_URL', 'https://ilinkai.weixin.qq.com').rstrip('/'),
        ))
        self.auth = AuthService(self.client, self.credentials)
        self.loop: MessageLoop | None = None
        self.loop_task = None
        self.reminder_task = None
        self.last_error = ''
        self.recent_senders: list[str] = []
        self.sender_contexts: dict[str, str] = {}

    async def start_login(self):
        return await self.auth.start()

    async def poll_login(self):
        result = await self.auth.poll()
        if result['status'] == 'confirmed':
            await self.start_loop()
        return result

    async def start_loop(self):
        if self.loop_task and not self.loop_task.done():
            return {'running': True}
        credentials = self.credentials.load()
        if not credentials:
            raise ILinkError('微信尚未完成扫码授权，或未设置 MEIMEI_WECHAT_BOT_TOKEN')
        self.client.set_credentials(credentials)
        self.loop = MessageLoop(self.client, self._handle_message)
        import asyncio
        self.loop_task = asyncio.create_task(self.loop.run())
        self.reminder_task = asyncio.create_task(self._reminder_loop())
        return {'running': True}

    async def stop(self):
        import asyncio
        if self.loop:
            self.loop.stop()
        if self.loop_task:
            try:
                await asyncio.wait_for(asyncio.shield(self.loop_task), timeout=2.0)
            except asyncio.TimeoutError:
                self.loop_task.cancel()
                try:
                    await self.loop_task
                except asyncio.CancelledError:
                    pass
            except Exception:
                pass
            self.loop_task = None
        if self.reminder_task:
            self.reminder_task.cancel()
            try:
                await self.reminder_task
            except asyncio.CancelledError:
                pass
            self.reminder_task = None
        await self.client.close()

    def status(self) -> dict[str, Any]:
        credentials = self.credentials.load()
        policy = public_config()
        session_expired = bool(self.loop and self.loop.session_expired)
        return {
            'configured': bool(credentials),
            'running': bool(self.loop_task and not self.loop_task.done()),
            'login': self.auth.status(),
            'last_error': self.loop.last_error if self.loop else self.last_error,
            'processed': self.loop.processed if self.loop else 0,
            'needs_relogin': session_expired,
            'account_id': credentials.account_id if credentials else '',
            'policy': policy,
            'recent_senders': list(self.recent_senders),
        }

    def policy(self) -> dict[str, Any]:
        return public_config()

    def update_policy(self, allow_users: list[str], allow_all: bool) -> dict[str, Any]:
        return save_config(allow_users, allow_all)

    async def _handle_message(self, message: IncomingText):
        self._remember_sender(message.from_user_id)
        self._remember_context(message.from_user_id, message.context_token)
        policy = public_config()
        if not policy['allow_all'] and message.from_user_id not in set(policy['allow_users']):
            await self.client.send_message(
                message.from_user_id,
                message.context_token,
                f'你尚未获得美美工作台的使用授权。请管理员将此用户 ID 加入白名单：{message.from_user_id}',
            )
            return
        command = message.text.strip()
        if command in {'/新会话', '/清空会话'}:
            SessionStore().clear(f'wechat:{message.from_user_id}')
            await self.client.send_message(
                message.from_user_id,
                message.context_token,
                '已清空当前对话上下文，凯凯小兵准备开始新的对话。',
            )
            return
        typing_ticket = await self._start_typing(message)
        runner = AgentRunner()
        try:
            answer = await runner.chat(
                f'wechat:{message.from_user_id}',
                message.text,
                channel='wechat',
                actor_id=message.from_user_id,
            )
        except ModelError as exc:
            answer = f'凯凯小兵暂时无法回答：{exc}'
        finally:
            await self._stop_typing(message, typing_ticket)
        await self.client.send_message(message.from_user_id, message.context_token, answer)

    async def _start_typing(self, message: IncomingText) -> str:
        try:
            config = await self.client.get_config(message.from_user_id, message.context_token)
            ticket = str(config.get('typing_ticket') or '')
            if ticket:
                await self.client.send_typing(message.from_user_id, ticket, 1)
            return ticket
        except ILinkError:
            return ''

    async def _stop_typing(self, message: IncomingText, ticket: str):
        if not ticket:
            return
        try:
            await self.client.send_typing(message.from_user_id, ticket, 2)
        except ILinkError:
            pass

    def _remember_sender(self, user_id: str):
        if not user_id:
            return
        self.recent_senders = [item for item in self.recent_senders if item != user_id]
        self.recent_senders.insert(0, user_id)
        del self.recent_senders[20:]

    async def _reminder_loop(self):
        while True:
            try:
                await self.send_pending_reminders()
                await asyncio.sleep(15 * 60)
            except asyncio.CancelledError:
                raise
            except Exception as exc:
                self.last_error = str(exc)
                await asyncio.sleep(60)

    async def send_pending_reminders(self) -> dict:
        """发送逾期/今日工作项摘要，并以任务、渠道用户和状态去重。"""
        policy = public_config()
        recipients = list(policy.get('allow_users') or []) if not policy.get('allow_all') else list(self.recent_senders)
        recipients = [item for item in recipients if item in self.sender_contexts]
        if not recipients:
            return {'sent': 0, 'skipped': 0, 'recipients': 0}
        items = work_items.list_work_items(bucket='overdue', reference_date=clock.today(), limit=30)
        items += work_items.list_work_items(bucket='today', reference_date=clock.today(), limit=30)
        unique = {int(item['id']): item for item in items}
        conn = db.get_conn()
        sent = skipped = 0
        for recipient in recipients:
            for item in unique.values():
                reminder_key = f"{item['status']}:{item.get('due_at') or item.get('scheduled_at') or ''}"
                exists = conn.execute(
                    '''SELECT 1 FROM wechat_reminder_receipts
                       WHERE class_id=? AND term_id=? AND task_id=? AND recipient=? AND reminder_key=?''',
                    (item['class_id'], item['term_id'], item['id'], recipient, reminder_key),
                ).fetchone()
                if exists:
                    skipped += 1
                    continue
                timing = item.get('timing_state') or '待处理'
                text = f"凯凯小兵提醒：{timing}有待处理事项——{item['title']}"
                web_url = os.environ.get('MEIMEI_WORKBENCH_WEB_URL', '')
                if web_url:
                    text += f"\n网页处理：{web_url.rstrip('/')}/#/tasks"
                await self.client.send_message(recipient, self.sender_contexts[recipient], text)
                conn.execute(
                    '''INSERT OR IGNORE INTO wechat_reminder_receipts(
                           class_id, term_id, task_id, recipient, reminder_key)
                       VALUES(?,?,?,?,?)''',
                    (item['class_id'], item['term_id'], item['id'], recipient, reminder_key),
                )
                sent += 1
        conn.commit()
        return {'sent': sent, 'skipped': skipped, 'recipients': len(recipients)}

    def _remember_context(self, user_id: str, context_token: str):
        if not hasattr(self, 'sender_contexts'):
            self.sender_contexts = {}
        if user_id and context_token:
            self.sender_contexts[user_id] = context_token


wechat_service = WeChatService()

File: server/src/wechat/service.ts
import { AgentRunner } from '../agent/runner.js';
import { SessionStore } from '../agent/sessionStore.js';
import { ModelError } from '../agent/modelClient.js';
import { getDb } from '../services/context.js';
import * as workItems from '../services/workItems.js';
import { AuthService } from './authService.js';
import { publicPolicy as configPublicPolicy, updatePolicy as savePolicy, loadConfig, type WeChatPolicy } from './config.js';
import { CredentialStore } from './credentialStore.js';
import { ILinkClient, ILinkError } from './ilinkClient.js';
import { MessageLoop } from './messageLoop.js';
import type { IncomingText } from './models.js';

const DEFAULT_ILINK_BASE_URL = 'https://ilinkai.weixin.qq.com';

export class WeChatService {
  credentials: CredentialStore;
  client: ILinkClient;
  auth: AuthService;
  loop: MessageLoop | null = null;
  loopTask: Promise<void> | null = null;
  reminderTask: Promise<void> | null = null;
  lastError = '';
  recentSenders: string[] = [];
  senderContexts: Record<string, string> = {};
  private reminderStopped = true;
  private reminderTimer: NodeJS.Timeout | null = null;

  constructor() {
    this.credentials = new CredentialStore();
    this.client = new ILinkClient({
      base_url: (process.env.MEIMEI_WECHAT_BASE_URL ?? DEFAULT_ILINK_BASE_URL).replace(/\/+$/, ''),
    });
    this.auth = new AuthService(this.client, this.credentials);
  }

  /** 登录/循环前按最新配置（env > DB agent_settings）刷新客户端；base_url 变化时重建。 */
  private syncConfig(): void {
    const config = loadConfig();
    const envBase = (process.env.MEIMEI_WECHAT_BASE_URL ?? '').replace(/\/+$/, '');
    const target = (envBase || config.base_url || DEFAULT_ILINK_BASE_URL).replace(/\/+$/, '');
    const current = (this.client as unknown as { config?: { base_url?: string } }).config?.base_url ?? '';
    if (current.replace(/\/+$/, '') !== target) {
      this.client = new ILinkClient({ base_url: target });
      this.auth = new AuthService(this.client, this.credentials);
    }
  }

  async startLogin(): Promise<Record<string, unknown>> {
    this.syncConfig();
    return this.auth.start();
  }

  async pollLogin(): Promise<Record<string, unknown>> {
    const result = await this.auth.poll();
    if (result.status === 'confirmed') {
      await this.startLoop();
    }
    return result;
  }

  async startLoop(): Promise<Record<string, unknown>> {
    this.syncConfig();
    if (this.loopTask) return { running: true };
    const credentials = this.credentials.load();
    if (!credentials) {
      throw new ILinkError('微信尚未完成扫码授权，或未设置 MEIMEI_WECHAT_BOT_TOKEN');
    }
    this.client.setCredentials(credentials);
    this.loop = new MessageLoop(this.client, (message) => this.handleMessage(message));
    const task = this.loop.run();
    this.loopTask = task;
    task.then(() => this.clearLoopTask(task), () => this.clearLoopTask(task));
    this.reminderStopped = false;
    this.reminderTask = this.runReminderLoop();
    return { running: true };
  }

  async stop(): Promise<void> {
    if (this.loop) this.loop.stop();
    if (this.loopTask) {
      this.client.abort();
      await Promise.race([this.loopTask, sleep(2000)]);
      this.loopTask = null;
    }
    this.reminderStopped = true;
    if (this.reminderTimer) {
      clearTimeout(this.reminderTimer);
      this.reminderTimer = null;
    }
    if (this.reminderTask) {
      await Promise.race([this.reminderTask, sleep(2000)]);
      this.reminderTask = null;
    }
    await this.client.close();
  }

  status(): Record<string, unknown> {
    const credentials = this.credentials.load();
    const policy = configPublicPolicy();
    const sessionExpired = Boolean(this.loop && this.loop.sessionExpired);
    return {
      configured: Boolean(credentials),
      running: Boolean(this.loopTask),
      login: this.auth.status(),
      last_error: this.loop ? this.loop.lastError : this.lastError,
      processed: this.loop ? this.loop.processed : 0,
      needs_relogin: sessionExpired,
      account_id: credentials ? credentials.account_id : '',
      policy,
      recent_senders: [...this.recentSenders],
    };
  }

  policy(): WeChatPolicy {
    return configPublicPolicy();
  }

  updatePolicy(allowUsers: string[], allowAll: boolean): WeChatPolicy {
    return savePolicy(allowUsers, allowAll);
  }  private async handleMessage(message: IncomingText): Promise<void> {
    this.rememberSender(message.from_user_id);
    this.rememberContext(message.from_user_id, message.context_token);
    const policy = configPublicPolicy();
    if (!policy.allow_all && !policy.allow_users.includes(message.from_user_id)) {
      await this.client.sendMessage(
        message.from_user_id,
        message.context_token,
        `你尚未获得美美工作台的使用授权。请管理员将此用户 ID 加入白名单：${message.from_user_id}`,
      );
      return;
    }
    const command = message.text.trim();
    if (command === '/新会话' || command === '/清空会话') {
      new SessionStore().clearOwned(
        `wechat:${message.from_user_id}`, message.from_user_id, 'wechat');
      await this.client.sendMessage(
        message.from_user_id,
        message.context_token,
        '已清空当前对话上下文，凯凯小兵准备开始新的对话。',
      );
      return;
    }
    const typingTicket = await this.startTyping(message);
    const runner = new AgentRunner();
    let answer: string;
    try {
      answer = await runner.chat(
        `wechat:${message.from_user_id}`,
        message.text,
        { channel: 'wechat', actorId: message.from_user_id },
      );
    } catch (error) {
      if (!(error instanceof ModelError)) throw error;
      answer = `凯凯小兵暂时无法回答：${error.message}`;
    } finally {
      await this.stopTyping(message, typingTicket);
    }
    await this.client.sendMessage(message.from_user_id, message.context_token, answer);
  }

  private async startTyping(message: IncomingText): Promise<string> {
    try {
      const config = await this.client.getConfig(message.from_user_id, message.context_token);
      const ticket = String(config.typing_ticket || '');
      if (ticket) {
        await this.client.sendTyping(message.from_user_id, ticket, 1);
      }
      return ticket;
    } catch (error) {
      if (error instanceof ILinkError) return '';
      throw error;
    }
  }

  private async stopTyping(message: IncomingText, ticket: string): Promise<void> {
    if (!ticket) return;
    try {
      await this.client.sendTyping(message.from_user_id, ticket, 2);
    } catch (error) {
      if (!(error instanceof ILinkError)) throw error;
    }
  }

  private rememberSender(userId: string): void {
    if (!userId) return;
    this.recentSenders = this.recentSenders.filter((item) => item !== userId);
    this.recentSenders.unshift(userId);
    this.recentSenders.length = Math.min(this.recentSenders.length, 20);
  }

  private rememberContext(userId: string, contextToken: string): void {
    if (userId && contextToken) {
      this.senderContexts[userId] = contextToken;
    }
  }

  private async runReminderLoop(): Promise<void> {
    while (!this.reminderStopped) {
      try {
        await this.sendPendingReminders();
        await this.reminderSleep(15 * 60 * 1000);
      } catch (error) {
        this.lastError = error instanceof Error ? error.message : String(error);
        await this.reminderSleep(60 * 1000);
      }
    }
  }

  async sendPendingReminders(): Promise<Record<string, number>> {
    const policy = configPublicPolicy();
    const recipients = policy.allow_all
      ? [...this.recentSenders]
      : [...policy.allow_users];
    const targets = recipients.filter((item) => item in this.senderContexts);
    if (targets.length === 0) return { sent: 0, skipped: 0, recipients: 0 };
    let items = workItems.listWorkItems({ bucket: 'overdue', limit: 30 });
    items = items.concat(workItems.listWorkItems({ bucket: 'today', limit: 30 }));
    const unique = new Map<number, Record<string, unknown>>();
    for (const item of items) unique.set(Number(item.id), item);
    const conn = getDb().connInstance;
    let sent = 0;
    let skipped = 0;
    for (const recipient of targets) {
      for (const item of unique.values()) {
        const reminderKey = `${String(item.status)}:${String(item.due_at || item.scheduled_at || '')}`;
        const exists = conn.prepare(
          'SELECT 1 FROM wechat_reminder_receipts '
          + 'WHERE class_id=? AND term_id=? AND task_id=? AND recipient=? AND reminder_key=?',
        ).get(item.class_id, item.term_id, item.id, recipient, reminderKey);
        if (exists) {
          skipped += 1;
          continue;
        }
        const timing = String(item.timing_state || '待处理');
        let text = `凯凯小兵提醒：${timing}有待处理事项——${String(item.title)}`;
        const webUrl = process.env.MEIMEI_WORKBENCH_WEB_URL ?? '';
        if (webUrl) {
          text += `\n网页处理：${webUrl.replace(/\/+$/, '')}/#/tasks`;
        }
        await this.client.sendMessage(recipient, this.senderContexts[recipient], text);
        conn.prepare(
          'INSERT OR IGNORE INTO wechat_reminder_receipts('
          + 'class_id, term_id, task_id, recipient, reminder_key) VALUES(?,?,?,?,?)',
        ).run(item.class_id, item.term_id, item.id, recipient, reminderKey);
        sent += 1;
      }
    }
    return { sent, skipped, recipients: targets.length };
  }

  private reminderSleep(ms: number): Promise<void> {
    return new Promise((resolve) => {
      this.reminderTimer = setTimeout(() => {
        this.reminderTimer = null;
        resolve();
      }, ms);
    });
  }

  private clearLoopTask(task: Promise<void>): void {
    if (this.loopTask === task) this.loopTask = null;
    /* 会话过期后消息循环已停止：继续发提醒只会重复失败，同步暂停提醒。 */
    if (this.loop?.sessionExpired && !this.reminderStopped) {
      this.reminderStopped = true;
      if (this.reminderTimer) {
        clearTimeout(this.reminderTimer);
        this.reminderTimer = null;
      }
      this.lastError = '微信会话已过期，提醒已暂停，请重新扫码登录后恢复。';
    }
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export const wechatService = new WeChatService();
